File: fsharp/parallel-letter-frequency/ParallelLetterFrequency.fs
module ParallelLetterFrequency

open System

module Map =
    let merge = Map.fold (fun acc key value -> Map.add key ((Map.tryFind key acc |> Option.defaultValue 0) + value) acc)

let private asyncCalculateFrequency text = async {
    return
        text
        |> Seq.filter Char.IsLetter
        |> Seq.countBy Char.ToLower
        |> Map.ofSeq
}

let frequency texts = 
    texts
    |> Seq.map asyncCalculateFrequency   
    |> Async.Parallel
    |> Async.RunSynchronously 
    |> Seq.fold Map.merge Map.empty

File: java/parallel-letter-frequency/src/main/java/ParallelLetterFrequency.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class ParallelLetterFrequency {
    private final String[] texts;

    ParallelLetterFrequency(String[] texts) {
        this.texts = texts;
    }

    Map<Character, Integer> countLetters() {
        return Arrays.stream(texts)
                .parallel()
                .map(this::countLetters)
                .reduce(new HashMap<>(), this::combine);
    }

    private Map<Character, Integer> countLetters(String text) {
        Map<Character, Integer> counts = new HashMap<>();

        for (var c: text.toCharArray()) {
            if (!Character.isLetter(c))
                continue;

            counts.merge(Character.toLowerCase(c), 1, Integer::sum);
        }

        return counts;
    }

    private Map<Character, Integer> combine(Map<Character, Integer> map, Map<Character, Integer> acc) {
        for (var entry: map.entrySet()) {
            acc.merge(entry.getKey(), entry.getValue(), Integer::sum);
        }

        return acc;
    }
}
